File: src/appleseed/renderer/meta/tests/test_entitymap.cpp

//
// This source file is part of appleseed.
// Visit https://appleseedhq.net/ for additional information and resources.
//
// This software is released under the MIT license.
//
// Copyright (c) 2010-2013 Francois Beaune, Jupiter Jazz Limited
// Copyright (c) 2014-2018 Francois Beaune, The appleseedhq Organization
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
//

// appleseed.renderer headers.
#include "renderer/modeling/entity/entity.h"
#include "renderer/modeling/entity/entitymap.h"
#include "renderer/utility/testutils.h"

// appleseed.foundation headers.
#include "foundation/memory/autoreleaseptr.h"
#include "foundation/utility/test.h"
#include "foundation/utility/uid.h"

using namespace foundation;
using namespace renderer;

TEST_SUITE(Renderer_Modeling_Entity_EntityMap)
{
    TEST_CASE(Insert_SetsParentPointerOnEntity)
    {
        auto_release_ptr<Entity> entity(new DummyEntity("entity"));
        const Entity* entity_ptr = entity.get();

        Entity* parent = (Entity*)0x123;
        EntityMap m(parent);
        m.insert(entity);

        EXPECT_EQ(parent, entity_ptr->get_parent());
    }

    TEST_CASE(Swap_GivenEntityMapWithOneItemAndAnotherEmptyEntityMap_MovesItemToOtherContainer)
    {
        EntityMap m1;
        m1.insert(auto_release_ptr<Entity>(new DummyEntity("entity")));

        EntityMap m2;
        m2.swap(m1);

        EXPECT_TRUE(m1.empty());
        EXPECT_FALSE(m2.empty());
    }

    TEST_CASE(Swap_FixesParentPointerOnEntities)
    {
        Entity* parent1 = (Entity*)0x123;
        EntityMap m1(parent1);
        m1.insert(auto_release_ptr<Entity>(new DummyEntity("entity1")));

        Entity* parent2 = (Entity*)0x456;
        EntityMap m2(parent2);
        m2.insert(auto_release_ptr<Entity>(new DummyEntity("entity2")));

        m2.swap(m1);

        EXPECT_EQ(parent1, m1.get_by_name("entity2")->get_parent());
        EXPECT_EQ(parent2, m2.get_by_name("entity1")->get_parent());
    }

    TEST_CASE(Remove_GivenUID_RemovesEntity)
    {
        auto_release_ptr<Entity> entity(new DummyEntity("entity"));
        const UniqueID entity_id = entity->get_uid();

        EntityMap m;
        m.insert(entity);

        m.remove(entity_id);

        EXPECT_TRUE(m.empty());
    }

    TEST_CASE(Remove_GivenUID_ReleasesEntity)
    {
        bool release_was_called = false;
        auto_release_ptr<Entity> entity(new DummyEntityReleaseCheck("entity", release_was_called));
        const UniqueID entity_id = entity->get_uid();

        EntityMap m;
        m.insert(entity);

        m.remove(entity_id);

        EXPECT_TRUE(release_was_called);
    }

    TEST_CASE(GetByUID_GivenUID_ReturnsEntity)
    {
        auto_release_ptr<Entity> entity1(new DummyEntity("entity1"));
        auto_release_ptr<Entity> entity2(new DummyEntity("entity2"));
        const UniqueID entity2_id = entity2->get_uid();
        const Entity* entity2_ptr = entity2.get();

        EntityMap m;
        m.insert(entity1);
        m.insert(entity2);

        EXPECT_EQ(entity2_ptr, m.get_by_uid(entity2_id));
    }

    TEST_CASE(GetByName_GivenName_ReturnsEntity)
    {
        auto_release_ptr<Entity> entity1(new DummyEntity("entity1"));
        auto_release_ptr<Entity> entity2(new DummyEntity("entity2"));
        const Entity* entity2_ptr = entity2.get();

        EntityMap m;
        m.insert(entity1);
        m.insert(entity2);

        EXPECT_EQ(entity2_ptr, m.get_by_name("entity2"));
    }
}

File: src/appleseed.python/test/testentitymap.py
#
# This source file is part of appleseed.
# Visit https://appleseedhq.net/ for additional information and resources.
#
# This software is released under the MIT license.
#
# Copyright (c) 2016-2018 Esteban Tovagliari, The appleseedhq Organization
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#

import unittest
import appleseed as asr


class TestEntityMap(unittest.TestCase):
    """
    Basic entity map tests.
    """

    def setUp(self):
        self.scn = asr.Scene()
        self.assembly_map = self.scn.assemblies()

    def test_get_by_name(self):
        ass = asr.Assembly("assembly", {})
        self.assembly_map.insert(ass)

        ass = asr.Assembly("another_assembly", {})
        self.assembly_map.insert(ass)

        a = self.assembly_map.get_by_name("assembly")
        self.assertEqual(a.get_name(), "assembly")

        a = self.assembly_map.get_by_name("another_assembly")
        self.assertEqual(a.get_name(), "another_assembly")

        a = self.assembly_map.get_by_name("no_such_assembly")
        self.assertEqual(a, None)

    def test_get_by_uuid(self):
        ass = asr.Assembly("assembly")
        uid1 = ass.get_uid()
        self.assembly_map.insert(ass)

        ass = asr.Assembly("another_assembly")
        uid2 = ass.get_uid()
        self.assembly_map.insert(ass)

        a = self.assembly_map.get_by_uid(uid1)
        self.assertEqual(a.get_name(), "assembly")

        a = self.assembly_map.get_by_uid(uid2)
        self.assertEqual(a.get_name(), "another_assembly")

        a = self.assembly_map.get_by_uid(77567)
        self.assertEqual(a, None)

    def test_get_item(self):
        ass = asr.Assembly("assembly")
        uid1 = ass.get_uid()
        self.assembly_map.insert(ass)

        ass = asr.Assembly("another_assembly")
        uid2 = ass.get_uid()
        self.assembly_map.insert(ass)

        self.assertEqual(self.assembly_map["assembly"].get_uid(), uid1)
        self.assertEqual(self.assembly_map["another_assembly"].get_uid(), uid2)

    def test_insert_remove_by_uid(self):
        ass = asr.Assembly("assembly")
        self.assembly_map.insert(ass)
        self.assertEqual(len(self.assembly_map), 1)

        a = self.assembly_map.get_by_name("assembly")
        ass = self.assembly_map.remove_by_uid(a.get_uid())
        self.assertEqual(len(self.assembly_map), 0)

        self.assembly_map.insert(ass)
        self.assertEqual(len(self.assembly_map), 1)

    def test_keys(self):
        self.assertEqual(list(self.assembly_map.keys()), [])

        ass = asr.Assembly("assembly")
        self.assembly_map.insert(ass)

        ass = asr.Assembly("another_assembly")
        self.assembly_map.insert(ass)

        self.assertEqual(list(self.assembly_map.keys()), ["assembly", "another_assembly"])

    def test_values(self):
        ass = asr.Assembly("assembly")
        uid1 = ass.get_uid()
        self.assembly_map.insert(ass)

        ass = asr.Assembly("another_assembly")
        uid2 = ass.get_uid()
        self.assembly_map.insert(ass)

        values = list(self.assembly_map.values())
        self.assertEqual(len(values), 2)
        self.assertEqual(values[0].get_uid(), uid1)
        self.assertEqual(values[1].get_uid(), uid2)

    def test_iters(self):
        names = ['assembly', 'assembly2', 'assembly3']
        uids = []

        for name in names:
            ass = asr.Assembly(name)
            uids.append(ass.get_uid())
            self.assembly_map.insert(ass)

        result_names = []
        result_uids = []
        for ass in self.assembly_map:
            result_names.append(ass)
            result_uids.append(self.assembly_map[ass].get_uid())

        self.assertEqual(sorted(names), sorted(result_names))
        self.assertEqual(sorted(uids), sorted(result_uids))

if __name__ == "__main__":
    unittest.main()

